node/src/p2p.rs

```diff
@@ -699,6 +713,26 @@ where
         }
     }
 
+    fn dial_bootnodes(&mut self) {
+        self.event_pub.send(NodeEvent::ConnectingToBootnodes);
+
+        for (peer_id, addrs) in &self.bootnodes {
+            let dial_opts = DialOpts::peer_id(*peer_id)
+                .addresses(addrs.clone())
+                // This is needed when addresses are provided otherwise
+                // our `kademlia::Behaviour` wrapper will not cononicalize them.
```

Comment:
```suggestion
                // This is needed when addresses are provided in  otherwise
                // our `kademlia::Behaviour` wrapper will not canonicalize them.
```
I'm not sure I understand this comment 100%. Doc mentions this allows extending addresses in https://docs.rs/libp2p/latest/libp2p/swarm/trait.NetworkBehaviour.html#method.handle_pending_outbound_connection, so this is about addresses coming _from_ kademlia that we might want to add?

Reply:
Kademlia discovers peers and their addresses, let's say, PeerA with AddressA1 and AddressA2. Then dials `DialOpts::peer_id(PeerA)` ([1]), this does not contain the addresses inside. The swarm gets the `DialOpts` and pass it to the behavior (which includes our Kademlia behavior) by calling `handle_pending_outbound_connection` ([2]). Kademlia then returns `[AddressA1, AddressA2]` from that call ([3]) and swarm uses them ([4]). It is a back and forth situation.

[1]: https://github.com/libp2p/rust-libp2p/blob/4a16b6181d36cd703066a227deff57c6ebe91f7c/protocols/kad/src/behaviour.rs#L629-L631
[2]: https://github.com/libp2p/rust-libp2p/blob/4a16b6181d36cd703066a227deff57c6ebe91f7c/swarm/src/lib.rs#L470-L477
[3]: https://github.com/libp2p/rust-libp2p/blob/4a16b6181d36cd703066a227deff57c6ebe91f7c/protocols/kad/src/behaviour.rs#L2188-L2220
[4]: https://github.com/libp2p/rust-libp2p/blob/4a16b6181d36cd703066a227deff57c6ebe91f7c/swarm/src/lib.rs#L480